modules/http/src/main/java/org/mule/module/http/internal/listener/HttpListenerRegistry.java

```diff
@@ -361,58 +391,97 @@ private int getPathPartsSize(String path)
         return path.split(SLASH, -1);
     }
 
-    public class PathMap
+    private boolean isCatchAllPath(String path)
     {
+        return WILDCARD_CHARACTER.equals(path);
+    }
 
+    /**
+     * Represents a URI path, which can be a parent to regular sub paths, a catch all sub path (/*) and a URI param path (/{param}).
+     * Request handler and matcher pairs (handlers for different methods, for example) can be added and removed from it. If all
+     * its handlers and children handlers' are removed, it will notify its parent the path itself can be removed. This means that
+     * the tree-like structure that results from binding paths together changes as handlers are created and disposed so special care
+     * needs to be taken regarding available paths.
+     */
+    public class Path
+    {
         List<RequestHandlerMatcherPair> requestHandlerMatcherPairs = new ArrayList<>();
 
-        private Map<String, PathMap> subPaths = new HashMap<>();
-        private PathMap catchAllPathMap;
-        private PathMap catchAllCurrentPathMap;
+        private String name;
+        private Path parent;
+        private Map<String, Path> subPaths = new HashMap<>();
+        private Path catchAll;
+        private Path catchAllUriParam;
 
-        public PathMap getCatchAllPathMap()
+        /**
+         * Makes this path self-aware within the path tree so that it can interact with its parent.
+         *
+         * @param name the name for this path
+         * @param parent the parent path
+         */
+        public void addIdentity(String name, Path parent) {
```

Comment:
I don't know if this has to be done in the constructor (there would be Paths without name in this case).